File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
## 북, 동, 하, 서 ( 시계방향 )
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

n, m = map(int, input().split())
r, c, d = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]

visited = [[0]*m for _ in range(n)]


visited[r][c] = 1
cnt = 1

while True:
    flag = 0           
    for _ in range(4): 
        d = (d+3) % 4  
        nx = r + dx[d]
        ny = c + dy[d]

        if 0 <= nx < n and 0 <= ny < m and arr[nx][ny] == 0:
            if visited[nx][ny] == 0:
                visited[nx][ny] = 1
                cnt += 1
                r = nx
                c = ny
                flag = 1        ## 청소 했다는 뜻
                break

    if flag == 0:               ## 위의 for문에 들어가지 못했을 때
        if arr[r-dx[d]][c-dy[d]] == 1:
            print(cnt)
            break
        else:
            r, c = r-dx[d], c-dy[d]

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {

	private static int n, m;
	private static int[][] map;
	private static int[][] visited;

	public static void main(String[] args) throws Exception {

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		String[] split = in.readLine().split(" ");

		n = Integer.parseInt(split[0]);
		m = Integer.parseInt(split[1]);

		map = new int[n][m];
		visited = new int[n][m];
		split = in.readLine().split(" ");
		int r = Integer.parseInt(split[0]);
		int c = Integer.parseInt(split[1]);
		int d = Integer.parseInt(split[2]);
		// 0: 북쪽, 1: 동쪽, 2: 남쪽, 3: 서쪽

		for (int i = 0; i < n; i++) {
			split = in.readLine().split(" ");
			for (int j = 0; j < m; j++) {
				map[i][j] = Integer.parseInt(split[j]);
			}
		}

		visited[r][c] = 1;
		move(r, c, d);
		System.out.println(cnt);

	}

	// 북동남서(상우하좌) -> 서남동북
	private static int[] dx = { -1, 0, 1, 0 };
	private static int[] dy = { 0, 1, 0, -1 };
	private static int cnt = 1;

	private static void move(int x, int y, int d) {

		while (true) {
			boolean moved = false;

			for (int i = 0; i < 4; i++) {

				d = (d + 3) % 4; // 반시계로 회전
				int nx = x + dx[d];
				int ny = y + dy[d];

				if (nx < 0 || ny < 0 || nx >= n || ny >= m) {
					continue;
				}

				// 빈 칸일때, 반시계 방향으로 회전 후 앞으로 전진
				if (map[nx][ny] == 0 && visited[nx][ny] == 0) {
					visited[nx][ny] = 1;
					moved = true;
					cnt++;
					x = nx;
					y = ny;
					break;
				}
			}

			if (!moved) { // 네 방향 모두 청소되어 있거나, 벽인 경우 후진
				if (map[x - dx[d]][y - dy[d]] == 1) {
					break;
				}

				else {
					x -= dx[d];
					y -= dy[d];
				}

			}
		}

	}

}